rfd/0156-notifications.md

```diff
@@ -0,0 +1,404 @@
+   These notifications will be stored in the database under the key `/notifications/global/<notification id>`.
+   They can be configured to target users based on one of the following matchers:
+   1. Permissions: Targets all users with the specified RBAC permission(s).
+   2. Role: Targets all users with the specified role(s).
+   3. All: Targets all users.
+
+Since user-specific notifications are stored by username, notifications for SSO users will persist between sessions.
+
+Notifications will have a default TTL of 30 days, after which they will be automatically deleted. In some cases, notifications may also automatically be deleted prior to their TTL, for
+example, a notification reminding a user to complete their periodic access list review will be deleted once the user does it.
+
+#### Types and Protos
+
+###### Notification Item
+
+These will be stored under `/notification/user/<username>/<notification id>`. They can also also be nested in a `GlobalNotification`.
+
+```protobuf
+// Notification represents a notification item.
+message Notification {
+   // ID is the unique identifier for the notification.
+   string ID = 1 [(gogoproto.jsontag) = "id"];
+   // Code is the unique notification code.
+   string Code = 2 [(gogoproto.jsontag) = "code"];
+   // Title is the title/header of the notification.
+   string Title = 3 [(gogoproto.jsontag) = "title"];
+   // Description is the text content of the notification.
+   string Description = 4 [(gogoproto.jsontag) = "description"];
+   // Created is when the notification was created, in UNIX time.
+   google.protobuf.Timestamp Created = 5 [(gogoproto.jsontag) = "created"];
+   // TTL is how long the notification should last for before automatically expiring. Default is 30 days.
+   int64 TTL = 6 [(gogoproto.jsontag) = "ttl,omitempty"];
+   // Metadata is optional additional metadata to be added to the notification.
+   google.protobuf.Struct Metadata = 7 [(gogoproto.jsontag) = "metadata,omitempty"];
+}
+```
+
+###### User Notification Preferences Item
+
+Each user will have one `UserNotificationStates` object which will keep track of the notifications they have seen or dismissed.
+
+These will be stored once per user in the database under `/notifications/states/<username>`
+
+```protobuf
+// UserNotificationStates represents a user's notification states. This is to keep track
+// of the notifications the user has seen and dismissed.
+message UserNotificationStates {
+   // ReadNotifications is a list of notification IDs that the user has already seen and will be marked as read.
+   repeated string ReadNotifications = 1 [(gogoproto.jsontag) = "read_notifications"];
+   // DismissedNotifications is a list of notification IDs that the user has chosen to hide. These will be filtered out when notifications are fetched.
+   repeated string DismissedNotifications = 2 [(gogoproto.jsontag) = "dismissed_notifications"];
+}
```

Comment:
Deleting expired notifications from the backend: This would be done during the filtering stage when fetching notifications, we would check if a notification that was fetched is supposed to be expired based on its `Created` date and `TTL`, and if so, delete it from the backend db entirely (this deletes it for everyone else too).

Deleting read and missed notifications from the user's notifications states: This would also be done during the filtering stage, and would involve checking if there are notification ID's stored in `DismissedNotifications` or `ReadNotifications` that refer to notifications which no longer exist.